### PR TITLE
Fix data race on owns_pyobj_ accesses with non-GIL protected threads

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75871
* __->__ #75563

Fixes https://github.com/pytorch/pytorch/issues/75529

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: c10/core/TensorImpl.h [hunk 1]
```diff
@@ -2282,11 +2287,11 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   }
 
   bool owns_pyobj() {
-    return owns_pyobj_;
+    return static_cast<uintptr_t>(pyobj_) & 1;
   }
 
   void set_owns_pyobj(bool b) {
-    owns_pyobj_ = b;
+    pyobj_ = static_cast<PyObject*>(static_cast<uintptr_t>(_unchecked_untagged_pyobj()) | b);
```
Comment (kumpera) [inline c10/core/TensorImpl.h:2294]: nit: AFAICT, there's no guarantee that converting a bool to an int will only produce the value `1` as any non-zero int converts to `true`.

Comment (swolchok) [reply c10/core/TensorImpl.h:2294]: " the type bool can be converted to int with the value false becoming ​0​ and true becoming 1. " per https://en.cppreference.com/w/cpp/language/implicit_conversion ; static_cast will hit the implicit conversion per https://en.cppreference.com/w/cpp/language/static_cast item (3) .

Looks like clang makes this work by always ensuring bools have the value 0 or 1 in the first place: https://godbolt.org/z/3W6EqEs4K

**Comment by swolchok** [comment]
re: build failures, I think you need reinterpret_cast in at least one direction of int-to-pointer conversion.

What's the data race this is fixing and how is it meant to fix it?

**Comment by swolchok** [comment]
(I expect to end up having to mention that this can only possibly fix a data race by assuming that loading a pointer-sized value is atomic and that *technically* you should use std::atomic)

**Comment by albanD** [comment]
> Because accesses to this field must take out Python GIL, so you've got a lock to serialize reads/writes on this field.

I'm a bit confused by that?
The one place where this flag is used today is in the destructor (https://github.com/pytorch/pytorch/blob/c2124f5c669be6afc504e8fd3d4d9f943ab46962/c10/core/TensorImpl.cpp#L336) where the GIL is definitely not held.
I understand that this argument is moot as there is no multithreading issues if you're already in the destructor. But we should have a clear comment to that effect next to the functions if we expect to GIL to be held when accessing the pyobj_ field.

**Review by swolchok — APPROVED** [review]
LGTM

**Comment by ezyang** [comment]
@albanD it already has the comment that it's GIL protected

```
  // This field contains a weak reference to a PyObject representing
  // this Tensor.  It MUST NOT be a strong reference, as that would
  // create a reference cycle between Tensor and the PyObject.  If
  // pyobj is nullptr, when we transfer Tensor to Python, we allocate
  // a new PyObject for it and set this field.  This field does not
  // have to be protected by an atomic as it is only allowed to be
  // accessed when you hold the GIL.
```

**Pull request closed without merging**